lib/include/jxl/decode.h

```diff
@@ -418,6 +424,22 @@ JxlDecoderSetKeepOrientation(JxlDecoder* dec, JXL_BOOL keep_orientation);
 JXL_EXPORT JxlDecoderStatus
 JxlDecoderSetRenderSpotcolors(JxlDecoder* dec, JXL_BOOL render_spotcolors);
 
+/** Enables or disables coalescing of zero-duration frames. By default, frames
+ * are returned with coalescing enabled, i.e. all frames have the image
+ * dimensions, and are blended if needed. When coalescing is disabled, frames
+ * can have arbitrary dimensions, a non-zero crop offset, and blending is not
+ * performed. For display, coalescing is recommended. For loading a multi-layer
+ * still image as separate layers (as opposed to the merged image), coalescing
+ * has to be disabled.
+ *
+ * @param dec decoder object
+ * @param coalescing JXL_TRUE to enable coalescing (default), JXL_FALSE to
+ * disable it.
+ * @return JXL_DEC_SUCCESS if no error, JXL_DEC_ERROR otherwise.
+ */
+JXL_EXPORT JxlDecoderStatus JxlDecoderSetCoalescing(JxlDecoder* dec,
```

Comment:
It is an interesting point that it might be desirable that some of the extra channels are semantically "internal only" in a hypothetical encoder (which does not yet exist, to be clear). However, that is a problem that is completely orthogonal to coalescing or not: also in the current (always-coalescing) API, there is at the moment no way to differentiate between "internal" extra channels and "to be exposed" extra channels.

A potential solution for that problem could be to reserve a specific channel name to mean "internal only", and then in the normal (coalescing) case such channels would be hidden by the api. In the non-coalescing case, if we want to allow applications to do their own blending in all cases, then such "internal only" extra channels would have to be exposed.

I think for applications it is hard and error-prone to actually do their own blending in all cases, since the way frame blending works in jxl is very flexible (the source frame can be chosen, the alpha channel can be chosen, all extra channels can have their own blend mode, etc). I think it might be better to only support the basic Gimp/Photoshop use case of a linear chain of layers in the non-coalescing decode api, and return a decode error if more exotic stuff is happening (i.e. something created by a hypothetical future encoder), effectively forcing the application to try again with coalescing enabled (letting the decoder figure out the exotic stuff).

Reply:
For the "internal only" extra channels, probably it would be good to define a new ExtraChannelType for it (we have values 7-14 still available and reserved for future ec types, with values 15 and 16 (`kNonOptional` and `kOptional`) as generic catch-all values for future ec types - where the idea is that then the ec name will have to indicate the real type). We could add a channel type `kInternalOnly`, which like the frame type `kReferenceOnly` indicates that it is something that shouldn't be exposed in the API since it is only an auxiliary thing used for better compression, not a semantic element of the actual image/animation.

I would then argue that we should specify that `kInternalOnly` channels can only be used for blending in Patches, not for frame blending. That way, we can ensure that `kRegularFrame` frames can always be exposed as semantic elements of a multi-layer image/animation, while the API can still hide `kInternalOnly` channels and `kReferenceOnly` frames.

This is all mostly orthogonal to the question of SetCoalescing though.

Also: yes, adding new features has a cost and can introduce bugs. I do think this is an important feature though, and I think it's better to introduce it sooner than later, because otherwise we might end up making design choices that will make it nearly impossible to still add this feature without breaking existing applications.

Reply:
This problem is not orthogonal. The plan was to not expose these "internal" extra channels. The fact that you now need to use these other extra channels for blending is part of setting this value to false, not an orthogonal problem, so this should be addressed here. This new feature is already introducing a bug that is not being considered here, it is not a hypothetical. This PR needs to explain how these "internal" extra channels are handled with the default (true) and how that changes with the false value.

For example, an alternative implementation of this would be to add another function to get "internal" data. Right now you get the FrameHeader, but a different implementation of this feature (which can be done in the future without API design changes) could let you get the internal frame header info (number of internal extra channels used for blending, blending modes, etc).